src/Core/Managed/Shared/Extensibility/TelemetrySink.cs

```diff
@@ -0,0 +1,149 @@
+﻿namespace Microsoft.ApplicationInsights.Extensibility
+{
+    using System;
+    using System.Diagnostics;
+    using System.Threading;
+    using Microsoft.ApplicationInsights.Channel;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+
+    /// <summary>
+    /// Represents a destination for telemetry, consisting of a set of telemetry processors and a channel.
+    /// </summary>
+    public sealed class TelemetrySink : IDisposable, ITelemetryProcessor
+    {
+        private TelemetryConfiguration telemetryConfiguration;
+        private ITelemetryChannel telemetryChannel;
+        private bool shouldDisposeChannel;
+        private TelemetryProcessorChain telemetryProcessorChain;
+        private TelemetryProcessorChainBuilder telemetryProcessorChainBuilder;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="TelemetrySink"/> class.
+        /// </summary>
+        /// <param name="telemetryConfiguration">Telemetry configuration to use for the new <see cref="TelemetrySink"/> instance.</param>
+        /// <param name="telemetryChannel">Telemetry channel to use for the new <see cref="TelemetrySink"/> instance.</param>
+        public TelemetrySink(TelemetryConfiguration telemetryConfiguration, ITelemetryChannel telemetryChannel = null)
+        {
+            if (telemetryConfiguration == null)
+            {
+                throw new ArgumentNullException(nameof(telemetryConfiguration));
+            }
+
+            this.telemetryConfiguration = telemetryConfiguration;
+
+            if (telemetryChannel != null)
+            {
+                this.telemetryChannel = telemetryChannel;
+                this.shouldDisposeChannel = false;
+            }
+            else
+            {
+                this.telemetryChannel = new InMemoryChannel();
+                this.shouldDisposeChannel = true;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets an instance of the <see cref="TelemetryProcessorChainBuilder"/> that this sink is using.
+        /// </summary>
+        public TelemetryProcessorChainBuilder TelemetryProcessorChainBuilder
+        {
+            get
+            {
+                LazyInitializer.EnsureInitialized(ref this.telemetryProcessorChainBuilder, () => new TelemetryProcessorChainBuilder(this.telemetryConfiguration, this));
```

Comment:
Note that `EnsureInitialized` returns the value too. So this can be a one-liner:
`return LazyInitializer.EnsureInitialized(ref ...)`